### PR TITLE
refactor(write): modify Write API to match new protobuf definitions

### PR DESCRIPTION
**Pull request by jon-whit**
## Review Checklist
- [ ] I have clicked on ["allow edits by maintainers"](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [ ] I have added documentation for new/changed functionality in this PR or in a PR to [openfga.dev](https://github.com/openfga/openfga.dev) [Provide a link to any relevant PRs in the references section above]
- [ ] The correct base branch is being used, if not `main`
- [ ] I have added tests to validate that the change in functionality is working as expected

### CODE REVIEW

File: pkg/storage/storage.go [hunk 1]
```diff
@@ -35,7 +35,7 @@ func NewPaginationOptions(ps int32, contToken string) PaginationOptions {
 
 // Writes and Deletes are typesafe aliases for Write arguments.
 type Writes = []*openfgav1.TupleKey
-type Deletes = []*openfgav1.TupleKey
```
Comment (jon-whit) [inline pkg/storage/storage.go:38]: This update highlights that our storage interface is a bit to coupled with our protobufs. Ideally we'd take some time to refactor our storage interface so that it’s more agnostic of the protobuf layer. We can make better use of interfaces and abstract storage filter structs there to avoid protobuf impacts.

File: pkg/storage/sqlcommon/sqlcommon.go [hunk 2]
```diff
@@ -311,11 +311,6 @@ func Write(
 		id := ulid.MustNew(ulid.Timestamp(now), ulid.DefaultEntropy()).String()
 		objectType, objectID := tupleUtils.SplitObject(tk.GetObject())
 
-		conditionName, conditionContext, err := marshalRelationshipCondition(tk.GetCondition())
```
Comment (jon-whit) [inline pkg/storage/sqlcommon/sqlcommon.go:314]: - [ ] todo: add test for NULL on condition and condition context when a tuple is deleted from the store.

File: pkg/storage/sqlcommon/sqlcommon.go [hunk 3]
```diff
@@ -346,7 +341,7 @@ func Write(
 		changelogBuilder = changelogBuilder.Values(
 			store, objectType, objectID,
 			tk.GetRelation(), tk.GetUser(),
-			conditionName, conditionContext,
+			"", nil,
```
[the remark this answers is not in the document]
Comment (jon-whit) [reply pkg/storage/sqlcommon/sqlcommon.go]: Added

**Review by jpadilla — APPROVED** [review]

File: pkg/tuple/tuple.go [hunk 4]
```diff
@@ -10,6 +10,18 @@ import (
...
+type TupleWithCondition interface {
+	TupleWithoutCondition
+	GetCondition() *openfgav1.RelationshipCondition
+}
+
+type TupleWithoutCondition interface {
+	GetUser() string
+	GetObject() string
+	GetRelation() string
+	String() string
```
Comment (miparnisari) [inline pkg/tuple/tuple.go:22]: I don't think you should put `String` here.  I haven't tested it but i think`TupleWithCondition.String()` will delegate to `TupleWithoutCondition.String()`

Comment (jon-whit) [reply pkg/tuple/tuple.go:22]: That's intended. If you're composing tuples a TupleWithCondition is a TupleWithoutCondition + the GetCondition bit. So the TupleWithoutCondition is the base triplet and we have a string method for it. See it's usage. It's only used where we print `(object, relation, user)`, and its default implementation comes from the protobuf generated `String` method fo the `TupleKeyWithoutCondition`.

**Review by miparnisari — APPROVED** [review]
